compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp

```diff
@@ -35,20 +43,31 @@ bool needToPackSubByteElementBitWidth(unsigned bitWidth) {
   return bitWidth < 8 && llvm::isPowerOf2_32(bitWidth) && bitWidth != 1;
 }
 
+bool needToPackSubByteElementBitWidth(unsigned bitWidth) {
+  return needToPackSubByteElementBitWidthImpl(
+      bitWidth, /*isPackedStorage=*/clEnableI1Support);
+}
+
 bool needToPackSubByteElements(RankedTensorType shapedType) {
   unsigned bitWidth = IREE::Util::getTypeBitWidth(shapedType.getElementType());
-  return needToPackSubByteElementBitWidth(bitWidth);
+  // Two paths to enable packed storage for i1 tensors: the attribute or cl
+  // option. The cl option will be dropped once frontend supports emitting
+  // tensors with attributes.
+  bool isPackedStorage =
```

Comment:
We shouldn't be doing boolean checks on the encoding, but asking the encoding attr what the size is - that's why we have it. None of this file should ever know about packed encodings, just that there's an encoding attr and it has a method that allows computing the bit width.

Concretely: EncodingAttr should have a `getAlignedStorageElementType` or some other methods that defaults to what legalizeStorageElementTypeImpl is doing (`legalizeStorageElementType` has always bothered me, it doesn't mean what it thinks it means!).

The flag and all checks for isPackedStorage need to be removed from this file. `needToPackSubByteElementBitWidth` was always a hack and the it was promised when it was added that it would be cleaned up - and yet here we are :)

Reply:
Point taken. Creating an issue to track it: https://github.com/iree-org/iree/issues/19618